### PR TITLE
Add the DataAvailabilityHeader to BlockMeta

### LINKED ISSUE #373: Save and Load Block Data via IPFS
The next step to completing #184 is to save and load block data using IPFS. 

It's important to note the issue described in this comment https://github.com/lazyledger/lazyledger-core/issues/355#issuecomment-848743870 , where we need to at least be explicit, but preferably be able to configure, when data is being saved to IPFS as there is significant overhead.

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

This super simple PR adds the `DataAvailabilityHeader` to `BlockMeta`. It's required for #374, which will close #373

### CODE REVIEW

File: proto/tendermint/types/types.proto [hunk 1]
```diff
@@ -213,6 +213,7 @@ message BlockMeta {
   int64   block_size = 2;
   Header  header     = 3 [(gogoproto.nullable) = false];
   int64   num_txs    = 4;
+  DataAvailabilityHeader da_header = 5 [(gogoproto.nullable) = false];
```
Comment (evan-forbes) [inline proto/tendermint/types/types.proto:216]: @liamsi, what was the thinking behind not using `[(gogoproto.nullable) = false]` for the DAH in the `tmproto.Block`? I used it here, but can revert.

Comment (liamsi) [reply proto/tendermint/types/types.proto:216]: No particular reason. I want to use these gogoprote annotations as little as possible. Some of them break with the actual protobuf specification. 

@marbar3778 is there a good rule of thumb when to use and when to use them?

**Review by liamsi — APPROVED** [review]
👏🏼

File: proto/tendermint/types/types.proto (hunk 1 above)
Comment (adlerjohn) [reply proto/tendermint/types/types.proto:216]: Revisiting John's favorite topic, but the [deterministic protobuf spec](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-027-deterministic-protobuf-serialization.md) doesn't include any provisions for gogoproto annotations. In fact, if we stick to deterministic protobuf then `[(gogoproto.nullable) = false]` would be invalid.

Comment (Wondertan) [reply proto/tendermint/types/types.proto:216]: @adlerjohn, why do you think it is unmaintained?

File: types/block_meta.go [hunk 2]
```diff
@@ -10,10 +10,11 @@ import (
...
 type BlockMeta struct {
-	BlockID   BlockID `json:"block_id"`
-	BlockSize int     `json:"block_size"`
-	Header    Header  `json:"header"`
-	NumTxs    int     `json:"num_txs"`
+	BlockID   BlockID                `json:"block_id"`
+	BlockSize int                    `json:"block_size"`
+	Header    Header                 `json:"header"`
+	NumTxs    int                    `json:"num_txs"`
+	DAHeader  DataAvailabilityHeader `json:"da_header"`
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply types/block_meta.go:17]: good question. my thinking was that we shouldn't be editing it, the dah consists of three slices which should be passed as by value, the general rule of thumb to not use pointers unless it's necessary, and that's what [`Block`](https://github.com/lazyledger/lazyledger-core/blob/3e3fa306441421fe78f96400f4737dcf899fdb93/types/block.go#L161-L161) uses

this was also why I was using gogoproto to do the same thing for the proto types, but that was revereted here https://github.com/lazyledger/lazyledger-core/pull/372/commits/3e3fa306441421fe78f96400f4737dcf899fdb93

File: proto/tendermint/types/types.proto (hunk 1 above)
Comment (evan-forbes) [reply proto/tendermint/types/types.proto]: switched to not using gogo in https://github.com/lazyledger/lazyledger-core/pull/372/commits/3e3fa306441421fe78f96400f4737dcf899fdb93

File: types/block_meta.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Wondertan) [reply types/block_meta.go:17]: Ok, mostly my concern is to be consistent mostly about the use of a particular structure. For the case of DAHeader we don't do that and I think we need to change the places where we use pointers to values.

**Review by Wondertan — APPROVED** [review]

**Pull request merged**